zerver/lib/markdown/__init__.py

```diff
@@ -1710,10 +1720,27 @@ def clean(self, line: str) -> str:
         line = mention.MENTIONS_RE.sub(lambda m: "@_**{}**".format(m.group("match")), line)
         # Silence all the user group mentions inside blockquotes
         line = mention.USER_GROUP_MENTIONS_RE.sub(lambda m: "@_*{}*".format(m.group("match")), line)
+        # Don't convert emoticons present inside blockquotes.
+        # When a user, who has `Convert emoticons before sending` enabled, quotes the message
+        # containing emoticon, sent by another user, who has the same setting disabled, creates
+        # a confussion for the latter user as they see those as the converted emoticon.
+        line = self.COMPILED_EMOTICON_RE.sub(
+            lambda m: self.escape_emoticon(m.group("emoticon")), line
+        )
 
         # And then run the upstream processor's code for removing the '>'
         return super().clean(line)
 
+    def escape_emoticon(self, emoticon: str) -> str:
+        # We substitute the found emoticon with a placeholder that will not be processed
+        # further and converted back to the original emoticon by `UnescapePostprocessor`.
+        return "{STX}{ord0}{ETX}{STX}{ord1}{ETX}".format(
+            STX=markdown.util.STX,
+            ord0=ord(emoticon[0]),
+            ord1=ord(emoticon[1]),
+            ETX=markdown.util.ETX,
```

Comment:
Any better suggestions for the variable names `ord0` and `ord1`, or they are fine?

Reply:
I don't understand this code, so we could definitely do better.  Is the idea that these are the two characters of an emoticon?  What if one is made up of 3 characters?

Reply:
But we have all emoticons of 2 characters, right?

Reply:
We could easily add one that's `:-p` in the future, so we shouldn't bake in that assumption if we can avoid it.